### PR TITLE
#119

### PR DESCRIPTION
**Pull request by larubbio**
Add LanguageLocation class and combine language, location and circle modules into the region module

### CODE REVIEW

**Comment by frankhuster** [comment]
test this please

**Comment by larubbio** [comment]
it will fail. :(

**Comment by frankhuster** [comment]
I wanted to see what the results of the build were, the prior build had been erased...

**Comment by larubbio** [comment]
This is a live look in at the server room:

**Comment by frankhuster** [comment]
eeek

**Comment by larubbio** [comment]
I think this should build and pass now.  I had to do a ton of merging from @frankhuster code.  I then modified the datamodel to work around the cascade issue.  I then got everything working and was about to submit but .... I had to merge more of @frankhuster dang code.

Once the cascade issue is fixed the datamodel should be reverted.  Ideally instead of LangaugeLocation having a boolean defaultForCircle field, the Circle would point to it's default LanguageLocation.  This way having only one default is enforced in the datamodel.

cc @llavoie @koshalt @frankhuster

File: region/src/main/java/org/motechproject/nms/region/domain/validation/VillageValidator.java [hunk 1]
```diff
@@ -15,10 +15,17 @@ public void initialize(ValidVillage validVillage) {
 
     @Override
     public boolean isValid(Village village, ConstraintValidatorContext constraintValidatorContext) {
+        if (village == null) {
+            return true;
+        }
+
         if (village.getVcode() == null && village.getSvid() == null) {
+            constraintValidatorContext
```
Comment (frankhuster) [inline region/src/main/java/org/motechproject/nms/region/domain/validation/VillageValidator.java]: Sweet!

Comment (larubbio) [inline region/src/main/java/org/motechproject/nms/region/domain/validation/VillageValidator.java]: Except it currently throws an npe.  I put a ticket in platform for it.

File: region/src/main/java/org/motechproject/nms/region/service/impl/LanguageLocationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Autowired
+    public LanguageLocationServiceImpl(LanguageLocationDataService languageLocationDataService) {
+        this.languageLocationDataService = languageLocationDataService;
+    }
+
+    /**
+     * Given a languageLocationCode returns the object if one exists
+     * @param code the code for the language location record
+     * @return The languageLocationRecord if it exists
```
Comment (frankhuster) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LanguageLocationServiceImpl.java]: LanguageLocation object

**Comment by frankhuster** [comment]
So as of now, we're not recording history for Circle & all?
If we're cool with that, then lgtm.
Otherwise, we can fix in a subsequent PR.

**Comment by koshalt** [comment]
Should we create a tracking item to audit history tracking in the next
milestone?

**Comment by larubbio** [comment]
There was another issue with platform around history tracking (the stack
overflow).  I put an issue in for it.

**Comment by larubbio** [comment]
I reopened the history tracking issues and moved them to sprint 3

**Pull request merged**